nexus/src/db/queries/network_interface.rs

```diff
@@ -642,6 +256,152 @@ fn decode_database_error(
+            // loop, which, while it may not matter much, can be nearly
+            // trivially avoided by converting to u128, adding, and converting
+            // back. Given that these spaces can be _really_ big, that is
+            // probably worth doing.
+            network
+                .iter()
+                .nth(NUM_INITIAL_RESERVED_IP_ADDRESSES as _)
+                .unwrap_or_else(|| {
+                    panic!("Unexpectedly small IPv4 subnetwork: '{}'", network)
+                })
+                .into()
+        }
+    }
+}
+
+/// The `NextGuestIpv4Address` query is a `NextItem` query for choosing the next
+/// available IPv4 address for a guest interface.
+#[derive(Debug, Clone, Copy)]
+pub struct NextGuestIpv4Address {
+    inner: NextItem<
+        db::schema::network_interface::table,
+        IpNetwork,
+        db::schema::network_interface::dsl::ip,
+        Uuid,
+        db::schema::network_interface::dsl::subnet_id,
+    >,
+}
+
+impl NextGuestIpv4Address {
+    pub fn new(subnet: Ipv4Network, subnet_id: Uuid) -> Self {
+        let subnet = IpNetwork::from(subnet);
+        let net = IpNetwork::from(first_available_address(&subnet));
+        let max_offset = last_address_offset(&subnet);
+        Self { inner: NextItem::new_scoped(net, subnet_id, max_offset) }
+    }
+}
+
+delegate_query_fragment_impl!(NextGuestIpv4Address);
+
+/// A `[NextItem`] subquery that selects the next empty slot for an interface.
+///
+/// Instances are limited to 8 interfaces (RFD 135). This pushes a subquery that
+/// looks like:
+///
+/// ```sql
+/// SELECT COALESCE((
+///     SELECT
+///         next_slot
+///     FROM
+///         generate_series(0, <max nics per instance>)
+///     AS
+///         next_slot
+///     LEFT OUTER JOIN
+///         network_interface
+///     ON
+///         (instance_id, time_deleted IS NULL, slot) =
+///         (<instance_id>, TRUE, next_slot)
+///     WHERE
+///         slot IS NULL
+///     LIMIT 1)
+/// ), 0)
+/// ```
+///
+/// That is, we select the lowest slot that has not yet been claimed by an
+/// interface on this instance, or zero if there is no such instance at all.
+///
+/// Errors
+/// ------
+///
+/// Note that the `generate_series` function is inclusive of its upper bound.
+/// We intentionally use the upper bound of the maximum number of NICs per
+/// instance. In the case where there are no available slots (the current max
+/// slot number is 7), this query will return 8. However, this violates the
+/// check on the slot column being between `[0, 8)`. This check violation is
+/// used to detect the case when there are no slots available.
+#[derive(Debug, Clone, Copy)]
+pub struct NextNicSlot {
+    inner: NextItem<
+        db::schema::network_interface::table,
+        i16,
+        db::schema::network_interface::dsl::slot,
+        Uuid,
+        db::schema::network_interface::dsl::instance_id,
+    >,
```

Comment:
Nice, this looks really compact for an allocation query!

Reply:
Yeah, this is the meat of it. I'm hoping that we can just delegate almost everything to `NextItem`, with the right columns and search range types.